## 1. Credit card applications
<p>Commercial banks receive <em>a lot</em> of applications for credit cards. Many of them get rejected for many reasons, like high loan balances, low income levels, or too many inquiries on an individual's credit report, for example. Manually analyzing these applications is mundane, error-prone, and time-consuming (and time is money!). Luckily, this task can be automated with the power of machine learning and pretty much every commercial bank does so nowadays. In this notebook, we will build an automatic credit card approval predictor using machine learning techniques, just like the real banks do.</p>
<p><img src="https://assets.datacamp.com/production/project_558/img/credit_card.jpg" alt="Credit card being held in hand"></p>
<p>We'll use the <a href="http://archive.ics.uci.edu/ml/datasets/credit+approval">Credit Card Approval dataset</a> from the UCI Machine Learning Repository. The structure of this notebook is as follows:</p>
<ul>
<li>First, we will start off by loading and viewing the dataset.</li>
<li>We will see that the dataset has a mixture of both numerical and non-numerical features, that it contains values from different ranges, plus that it contains a number of missing entries.</li>
<li>We will have to preprocess the dataset to ensure the machine learning model we choose can make good predictions.</li>
<li>After our data is in good shape, we will do some exploratory data analysis to build our intuitions.</li>
<li>Finally, we will build a machine learning model that can predict if an individual's application for a credit card will be accepted.</li>
</ul>
<p>First, loading and viewing the dataset. We find that since this data is confidential, the contributor of the dataset has anonymized the feature names.</p>

In [6]:
# Import pandas
import pandas as pd

# Load dataset
cc_apps = pd.read_csv("datasets/cc_approvals.data",header=None)

# Inspect data
cc_apps.head(5)

0      1      2  3  4  5  6     7  8  9   10 11 12     13   14 15
0  b  30.83  0.000  u  g  w  v  1.25  t  t   1  f  g  00202    0  +
1  a  58.67  4.460  u  g  q  h  3.04  t  t   6  f  g  00043  560  +
2  a  24.50  0.500  u  g  q  h  1.50  t  f   0  f  g  00280  824  +
3  b  27.83  1.540  u  g  w  v  3.75  t  t   5  t  g  00100    3  +
4  b  20.17  5.625  u  g  w  v  1.71  t  f   0  f  s  00120    0  +

## 2. Inspecting the applications
<p>The output may appear a bit confusing at its first sight, but let's try to figure out the most important features of a credit card application. The features of this dataset have been anonymized to protect the privacy, but <a href="http://rstudio-pubs-static.s3.amazonaws.com/73039_9946de135c0a49daa7a0a9eda4a67a72.html">this blog</a> gives us a pretty good overview of the probable features. The probable features in a typical credit card application are <code>Gender</code>, <code>Age</code>, <code>Debt</code>, <code>Married</code>, <code>BankCustomer</code>, <code>EducationLevel</code>, <code>Ethnicity</code>, <code>YearsEmployed</code>, <code>PriorDefault</code>, <code>Employed</code>, <code>CreditScore</code>, <code>DriversLicense</code>, <code>Citizen</code>, <code>ZipCode</code>, <code>Income</code> and finally the <code>ApprovalStatus</code>. This gives us a pretty good starting point, and we can map these features with respect to the columns in the output.   </p>
<p>As we can see from our first glance at the data, the dataset has a mixture of numerical and non-numerical features. This can be fixed with some preprocessing, but before we do that, let's learn about the dataset a bit more to see if there are other dataset issues that need to be fixed.</p>

In [7]:
# Print summary statistics
cc_apps_description = cc_apps.describe()
print(cc_apps_description)

               2           7          10             14
count  690.000000  690.000000  690.00000     690.000000
mean     4.758725    2.223406    2.40000    1017.385507
std      4.978163    3.346513    4.86294    5210.102598
min      0.000000    0.000000    0.00000       0.000000
25%      1.000000    0.165000    0.00000       0.000000
50%      2.750000    1.000000    0.00000       5.000000
75%      7.207500    2.625000    3.00000     395.500000
max     28.000000   28.500000   67.00000  100000.000000


In [8]:
# Print DataFrame information
cc_apps_info = cc_apps.info()
print(cc_apps_info)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 690 entries, 0 to 689
Data columns (total 16 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   0       690 non-null    object 
 1   1       690 non-null    object 
 2   2       690 non-null    float64
 3   3       690 non-null    object 
 4   4       690 non-null    object 
 5   5       690 non-null    object 
 6   6       690 non-null    object 
 7   7       690 non-null    float64
 8   8       690 non-null    object 
 9   9       690 non-null    object 
 10  10      690 non-null    int64  
 11  11      690 non-null    object 
 12  12      690 non-null    object 
 13  13      690 non-null    object 
 14  14      690 non-null    int64  
 15  15      690 non-null    object 
dtypes: float64(2), int64(2), object(12)
memory usage: 86.4+ KB
None


In [9]:
# Inspect missing values in the dataset
cc_apps.tail(17).isnull().sum()

0     0
1     0
2     0
3     0
4     0
5     0
6     0
7     0
8     0
9     0
10    0
11    0
12    0
13    0
14    0
15    0
dtype: int64

In [10]:
def rename_columns(data):
    """
    Rename the columns of a DataFrame.

    Args:
        data (pd.DataFrame): The input DataFrame with original column names.

    Returns:
        pd.DataFrame: The DataFrame with columns renamed to specified names.

    Raises:
        ValueError: If the input is not a pandas DataFrame.
    """
    # Define the new column names as per your request
    new_column_names = [
        "Gender", "Age", "Debt", "Married", "BankCustomer", "EducationLevel",
        "Ethnicity", "YearsEmployed", "PriorDefault", "Employed", "CreditScore",
        "DriversLicense", "Citizen", "ZipCode", "Income", "ApprovalStatus"
    ]
    
    # Check if the input data is a pandas DataFrame
    if isinstance(data, pd.DataFrame):
        # Rename the columns of the DataFrame
        data.columns = new_column_names
        # Return the DataFrame with renamed columns
        return data
    else:
        # Raise an error if the input is not a DataFrame
        raise ValueError("Input must be a pandas DataFrame")

In [11]:
renamed_dataset = rename_columns(cc_apps)  # Rename the columns

In [12]:
renamed_dataset.head()

Gender    Age   Debt Married BankCustomer EducationLevel Ethnicity  \
0      b  30.83  0.000       u            g              w         v   
1      a  58.67  4.460       u            g              q         h   
2      a  24.50  0.500       u            g              q         h   
3      b  27.83  1.540       u            g              w         v   
4      b  20.17  5.625       u            g              w         v   

   YearsEmployed PriorDefault Employed  CreditScore DriversLicense Citizen  \
0           1.25            t        t            1              f       g   
1           3.04            t        t            6              f       g   
2           1.50            t        f            0              f       g   
3           3.75            t        t            5              t       g   
4           1.71            t        f            0              f       s   

  ZipCode  Income ApprovalStatus  
0   00202       0              +  
1   00043     560              +  
2   00280     824              +  
3   00100       3              +  
4   00120       0              +

In [14]:
cc_apps = renamed_dataset
cc_apps.head()

Gender    Age   Debt Married BankCustomer EducationLevel Ethnicity  \
0      b  30.83  0.000       u            g              w         v   
1      a  58.67  4.460       u            g              q         h   
2      a  24.50  0.500       u            g              q         h   
3      b  27.83  1.540       u            g              w         v   
4      b  20.17  5.625       u            g              w         v   

   YearsEmployed PriorDefault Employed  CreditScore DriversLicense Citizen  \
0           1.25            t        t            1              f       g   
1           3.04            t        t            6              f       g   
2           1.50            t        f            0              f       g   
3           3.75            t        t            5              t       g   
4           1.71            t        f            0              f       s   

  ZipCode  Income ApprovalStatus  
0   00202       0              +  
1   00043     560              +  
2   00280     824              +  
3   00100       3              +  
4   00120       0              +

## 3. Splitting the dataset into train and test sets
<p>Now, we will split our data into train set and test set to prepare our data for two different phases of machine learning modeling: training and testing. Ideally, no information from the test data should be used to preprocess the training data or should be used to direct the training process of a machine learning model. Hence, we first split the data and then preprocess it.</p>
<p>Also, features like <code>DriversLicense</code> and <code>ZipCode</code> are not as important as the other features in the dataset for predicting credit card approvals. To get a better sense, we can measure their <a href="https://realpython.com/numpy-scipy-pandas-correlation-python/">statistical correlation</a> to the labels of the dataset. But this is out of scope for this project. We should drop them to design our machine learning model with the best set of features. In Data Science literature, this is often referred to as <em>feature selection</em>. </p>

In [15]:
# Import train_test_split
from sklearn.model_selection import train_test_split

In [16]:
# Drop the features 11 and 13
cc_apps = cc_apps.drop(['DriversLicense','ZipCode'],axis=1)  # 11 --> DriverLiscense and 13 --> Zipcode

In [17]:
cc_apps.head()

Gender    Age   Debt Married BankCustomer EducationLevel Ethnicity  \
0      b  30.83  0.000       u            g              w         v   
1      a  58.67  4.460       u            g              q         h   
2      a  24.50  0.500       u            g              q         h   
3      b  27.83  1.540       u            g              w         v   
4      b  20.17  5.625       u            g              w         v   

   YearsEmployed PriorDefault Employed  CreditScore Citizen  Income  \
0           1.25            t        t            1       g       0   
1           3.04            t        t            6       g     560   
2           1.50            t        f            0       g     824   
3           3.75            t        t            5       g       3   
4           1.71            t        f            0       s       0   

  ApprovalStatus  
0              +  
1              +  
2              +  
3              +  
4              +

In [18]:
# Split into train and test sets
cc_apps_train, cc_apps_test = train_test_split(cc_apps, test_size=0.33, random_state=42)

## 4. Handling the missing values (part i)
<p>Now we've split our data, we can handle some of the issues we identified when inspecting the DataFrame, including:</p>
<ul>
<li>Our dataset contains both numeric and non-numeric data (specifically data that are of <code>float64</code>, <code>int64</code> and <code>object</code> types). Specifically, the features 2, 7, 10 and 14 contain numeric values (of types float64, float64, int64 and int64 respectively) and all the other features contain non-numeric values.</li>
<li>The dataset also contains values from several ranges. Some features have a value range of 0 - 28, some have a range of 2 - 67, and some have a range of 1017 - 100000. Apart from these, we can get useful statistical information (like <code>mean</code>, <code>max</code>, and <code>min</code>) about the features that have numerical values. </li>
<li>Finally, the dataset has missing values, which we'll take care of in this task. The missing values in the dataset are labeled with '?', which can be seen in the last cell's output of the second task.</li>
</ul>
<p>Now, let's temporarily replace these missing value question marks with NaN.</p>

In [20]:
cc_apps_train.tail()

Gender    Age   Debt Married BankCustomer EducationLevel Ethnicity  \
71       b  34.83  4.000       u            g              d        bb   
106      b  28.75  1.165       u            g              k         v   
270      b  37.58  0.000       ?            ?              ?         ?   
435      b  19.00  0.000       y            p             ff        ff   
102      b  18.67  5.000       u            g              q         v   

     YearsEmployed PriorDefault Employed  CreditScore Citizen  Income  \
71          12.500            t        f            0       g       0   
106          0.500            t        f            0       s       0   
270          0.000            f        f            0       p       0   
435          0.000            f        t            4       g       1   
102          0.375            t        t            2       g      38   

    ApprovalStatus  
71               -  
106              -  
270              +  
435              -  
102              -

In [22]:
cc_apps_test.tail(17)

Gender    Age    Debt Married BankCustomer EducationLevel Ethnicity  \
409      b  17.08   0.250       u            g              q         v   
588      b  26.67   1.750       y            p              c         v   
177      a  26.08   8.665       u            g             aa         v   
449      b  20.00   7.000       u            g              c         v   
656      b  25.67   3.250       u            g              c         h   
83       a      ?   3.500       u            g              d         v   
447      a  27.33   1.665       u            g             ff        ff   
349      b  27.83   1.500       u            g              w         v   
453      ?  29.75   0.665       u            g              w         v   
79       b  21.50   9.750       u            g              c         v   
23       a  27.42  14.500       u            g              x         h   
15       b  36.67   4.415       y            p              k         v   
375      a  20.83   0.500       y            p              e        dd   
234      a  58.42  21.000       u            g              i        bb   
644      b  36.17   0.420       y            p              w         v   
271      b  32.33   2.500       u            g              c         v   
311      b  19.00   1.750       y            p              c         v   

     YearsEmployed PriorDefault Employed  CreditScore Citizen  Income  \
409          0.335            f        t            4       g       8   
588          1.000            t        t            5       g    5777   
177          1.415            t        f            0       g     150   
449          0.500            f        f            0       g       0   
656          2.290            f        t            1       g      21   
83           3.000            t        f            0       g       0   
447          0.000            f        f            0       g       1   
349          2.250            f        t            1       g       3   
453          0.250            f        f            0       g       0   
79           0.250            t        f            0       g       0   
23           3.085            t        t            1       g      11   
15           0.250            t        t           10       g       0   
375          1.000            f        f            0       g       0   
234         10.000            t        t           13       g    6700   
644          0.290            f        f            0       g       2   
271          1.250            f        f            0       g       0   
311          2.335            f        f            0       g       6   

    ApprovalStatus  
409              -  
588              +  
177              +  
449              -  
656              -  
83               -  
447              -  
349              -  
453              -  
79               -  
23               +  
15               +  
375              -  
234              +  
644              -  
271              -  
311              -

In [23]:
# Import numpy
import numpy as np

# Replace the '?'s with NaN in the train and test sets
cc_apps_train = cc_apps_train.replace('?',np.nan)
cc_apps_test = cc_apps_test.replace('?',np.nan)

## 5. Handling the missing values (part ii)
<p>We replaced all the question marks with NaNs. This is going to help us in the next missing value treatment that we are going to perform.</p>
<p>An important question that gets raised here is <em>why are we giving so much importance to missing values</em>? Can't they be just ignored? Ignoring missing values can affect the performance of a machine learning model heavily. While ignoring the missing values our machine learning model may miss out on information about the dataset that may be useful for its training. Then, there are many models which cannot handle missing values implicitly such as Linear Discriminant Analysis (LDA). </p>
<p>So, to avoid this problem, we are going to impute the missing values with a strategy called mean imputation.</p>

In [25]:
# Impute the missing values with mean imputation
numeric_columns = ['Age', 'YearsEmployed', 'CreditScore', 'Income']
for col in numeric_columns:
    col_mean = cc_apps_train[col].astype(float).mean()
    cc_apps_train[col] = cc_apps_train[col].replace(np.nan, col_mean)
    cc_apps_test[col] = cc_apps_test[col].replace(np.nan, col_mean)

In [26]:
# Count the number of NaNs in the datasets and print the counts to verify
# print("Number of NaNs in cc_apps_train:")
print(cc_apps_train.isna().sum())

Gender            8
Age               0
Debt              0
Married           6
BankCustomer      6
EducationLevel    7
Ethnicity         7
YearsEmployed     0
PriorDefault      0
Employed          0
CreditScore       0
Citizen           0
Income            0
ApprovalStatus    0
dtype: int64


In [27]:
# print("\nNumber of NaNs in cc_apps_test:")
print(cc_apps_test.isna().sum())

Gender            4
Age               0
Debt              0
Married           0
BankCustomer      0
EducationLevel    2
Ethnicity         2
YearsEmployed     0
PriorDefault      0
Employed          0
CreditScore       0
Citizen           0
Income            0
ApprovalStatus    0
dtype: int64


## 6. Handling the missing values (part iii)
<p>We have successfully taken care of the missing values present in the numeric columns. There are still some missing values to be imputed for columns 0, 1, 3, 4, 5, 6 and 13. All of these columns contain non-numeric data and this is why the mean imputation strategy would not work here. This needs a different treatment. </p>
<p>We are going to impute these missing values with the most frequent values as present in the respective columns. This is <a href="https://www.datacamp.com/community/tutorials/categorical-data">good practice</a> when it comes to imputing missing values for categorical data in general.</p>

In [28]:
# Iterate over each column of cc_apps_train
for col in cc_apps_train.columns:
    # Check if the column is of object type
    if cc_apps_train[col].dtype == 'object':
        # Impute with the most frequent value
        cc_apps_train[col] = cc_apps_train[col].fillna(cc_apps_train[col].value_counts().index[0])
        cc_apps_test[col] = cc_apps_test[col].fillna(cc_apps_train[col].value_counts().index[0])

In [29]:
# Count the number of NaNs in the dataset and print the counts to verify
print("Number of NaNs in cc_apps_train:")
print(cc_apps_train.isna().sum())

Number of NaNs in cc_apps_train:
Gender            0
Age               0
Debt              0
Married           0
BankCustomer      0
EducationLevel    0
Ethnicity         0
YearsEmployed     0
PriorDefault      0
Employed          0
CreditScore       0
Citizen           0
Income            0
ApprovalStatus    0
dtype: int64


In [30]:
print("\nNumber of NaNs in cc_apps_test:")
print(cc_apps_test.isna().sum())


Number of NaNs in cc_apps_test:
Gender            0
Age               0
Debt              0
Married           0
BankCustomer      0
EducationLevel    0
Ethnicity         0
YearsEmployed     0
PriorDefault      0
Employed          0
CreditScore       0
Citizen           0
Income            0
ApprovalStatus    0
dtype: int64


In [31]:
cc_apps_train.head()

Gender    Age   Debt Married BankCustomer EducationLevel Ethnicity  \
382      a  24.33  2.500       y            p              i        bb   
137      b  33.58  2.750       u            g              m         v   
346      b  32.25  1.500       u            g              c         v   
326      b  30.17  1.085       y            p              c         v   
33       a  36.75  5.125       u            g              e         v   

     YearsEmployed PriorDefault Employed  CreditScore Citizen  Income  \
382           4.50            f        f            0       g     456   
137           4.25            t        t            6       g       0   
346           0.25            f        f            0       g     122   
326           0.04            f        f            0       g     179   
33            5.00            t        f            0       g    4000   

    ApprovalStatus  
382              -  
137              +  
346              -  
326              -  
33               +

## 7. Preprocessing the data (part i)
<p>The missing values are now successfully handled.</p>
<p>There is still some minor but essential data preprocessing needed before we proceed towards building our machine learning model. We are going to divide these remaining preprocessing steps into two main tasks:</p>
<ol>
<li>Convert the non-numeric data into numeric.</li>
<li>Scale the feature values to a uniform range.</li>
</ol>
<p>First, we will be converting all the non-numeric values into numeric ones. We do this because not only it results in a faster computation but also many machine learning models (like XGBoost) (and especially the ones developed using scikit-learn) require the data to be in a strictly numeric format. We will do this by using the <code>get_dummies()</code> method from pandas.</p>

In [35]:
# Convert the categorical features in the train and test sets independently
cc_apps_train = pd.get_dummies(cc_apps_train)
cc_apps_test = pd.get_dummies(cc_apps_test)

# Reindex the columns of the test set aligning with the train set
cc_apps_test = cc_apps_test.reindex(columns=cc_apps_train.columns, fill_value=0)


In [36]:
cc_apps_train.head()

Debt  YearsEmployed  CreditScore  Income  Gender_a  Gender_b  \
382  2.500           4.50            0     456      True     False   
137  2.750           4.25            6       0     False      True   
346  1.500           0.25            0     122     False      True   
326  1.085           0.04            0     179     False      True   
33   5.125           5.00            0    4000      True     False   

     Age_31.635754923413568  Age_13.75  Age_15.83  Age_15.92  ...  \
382                   False      False      False      False  ...   
137                   False      False      False      False  ...   
346                   False      False      False      False  ...   
326                   False      False      False      False  ...   
33                    False      False      False      False  ...   

     Ethnicity_z  PriorDefault_f  PriorDefault_t  Employed_f  Employed_t  \
382        False            True           False        True       False   
137        False           False            True       False        True   
346        False            True           False        True       False   
326        False            True           False        True       False   
33         False           False            True        True       False   

     Citizen_g  Citizen_p  Citizen_s  ApprovalStatus_+  ApprovalStatus_-  
382       True      False      False             False              True  
137       True      False      False              True             False  
346       True      False      False             False              True  
326       True      False      False             False              True  
33        True      False      False              True             False  

[5 rows x 330 columns]

## 8. Preprocessing the data (part ii)
<p>Now, we are only left with one final preprocessing step of scaling before we can fit a machine learning model to the data. </p>
<p>Now, let's try to understand what these scaled values mean in the real world. Let's use <code>CreditScore</code> as an example. The credit score of a person is their creditworthiness based on their credit history. The higher this number, the more financially trustworthy a person is considered to be. So, a <code>CreditScore</code> of 1 is the highest since we're rescaling all the values to the range of 0-1.</p>

In [38]:
# Import MinMaxScaler
from sklearn.preprocessing import MinMaxScaler

# Segregate features and labels into separate variables using iloc
X_train, y_train = cc_apps_train.iloc[:, :-1], cc_apps_train.iloc[:, -1]
X_test, y_test = cc_apps_test.iloc[:, :-1], cc_apps_test.iloc[:, -1]

# Instantiate MinMaxScaler and use it to rescale X_train and X_test
scaler = MinMaxScaler(feature_range=(0, 1))
rescaledX_train = scaler.fit_transform(X_train)
rescaledX_test = scaler.transform(X_test)


In [46]:
some = pd.DataFrame(rescaledX_train)
some.head()

0       1         2        3    4    5    6    7    8    9    ...  \
0  0.094931  0.2250  0.000000  0.00456  1.0  0.0  0.0  0.0  0.0  0.0  ...   
1  0.104424  0.2125  0.089552  0.00000  0.0  1.0  0.0  0.0  0.0  0.0  ...   
2  0.056958  0.0125  0.000000  0.00122  0.0  1.0  0.0  0.0  0.0  0.0  ...   
3  0.041200  0.0020  0.000000  0.00179  0.0  1.0  0.0  0.0  0.0  0.0  ...   
4  0.194608  0.2500  0.000000  0.04000  1.0  0.0  0.0  0.0  0.0  0.0  ...   

   319  320  321  322  323  324  325  326  327  328  
0  0.0  0.0  1.0  0.0  1.0  0.0  1.0  0.0  0.0  0.0  
1  1.0  0.0  0.0  1.0  0.0  1.0  1.0  0.0  0.0  1.0  
2  1.0  0.0  1.0  0.0  1.0  0.0  1.0  0.0  0.0  0.0  
3  1.0  0.0  1.0  0.0  1.0  0.0  1.0  0.0  0.0  0.0  
4  1.0  0.0  0.0  1.0  1.0  0.0  1.0  0.0  0.0  1.0  

[5 rows x 329 columns]

## 9. Fitting a logistic regression model to the train set
<p>Essentially, predicting if a credit card application will be approved or not is a <a href="https://en.wikipedia.org/wiki/Statistical_classification">classification</a> task. According to UCI, our dataset contains more instances that correspond to "Denied" status than instances corresponding to "Approved" status. Specifically, out of 690 instances, there are 383 (55.5%) applications that got denied and 307 (44.5%) applications that got approved. </p>
<p>This gives us a benchmark. A good machine learning model should be able to accurately predict the status of the applications with respect to these statistics.</p>
<p>Which model should we pick? A question to ask is: <em>are the features that affect the credit card approval decision process correlated with each other?</em> Although we can measure correlation, that is outside the scope of this notebook, so we'll rely on our intuition that they indeed are correlated for now. Because of this correlation, we'll take advantage of the fact that generalized linear models perform well in these cases. Let's start our machine learning modeling with a Logistic Regression model (a generalized linear model).</p>

In [47]:
# Import LogisticRegression
from sklearn.linear_model import LogisticRegression

# Instantiate a LogisticRegression classifier with default parameter values
logreg = LogisticRegression()

# Fit logreg to the train set
logreg.fit(rescaledX_train, y_train)


LogisticRegression()

## 10. Making predictions and evaluating performance
<p>But how well does our model perform? </p>
<p>We will now evaluate our model on the test set with respect to <a href="https://developers.google.com/machine-learning/crash-course/classification/accuracy">classification accuracy</a>. But we will also take a look the model's <a href="http://www.dataschool.io/simple-guide-to-confusion-matrix-terminology/">confusion matrix</a>. In the case of predicting credit card applications, it is important to see if our machine learning model is equally capable of predicting approved and denied status, in line with the frequency of these labels in our original dataset. If our model is not performing well in this aspect, then it might end up approving the application that should have been approved. The confusion matrix helps us to view our model's performance from these aspects.  </p>

In [50]:
# Import confusion_matrix
from sklearn.metrics import confusion_matrix

# Use logreg to predict instances from the test set and store it
y_pred = logreg.predict(rescaledX_test)

In [51]:
# Get the accuracy score of logreg model and print it
accuracy = logreg.score(rescaledX_test, y_test)
print("Accuracy of logistic regression classifier:", accuracy)

# Print the confusion matrix of the logreg model
conf_matrix = confusion_matrix(y_test, y_pred)
print("Confusion Matrix:\n", conf_matrix)


Accuracy of logistic regression classifier: 1.0
Confusion Matrix:
 [[103   0]
 [  0 125]]


## 11. Grid searching and making the model perform better
<p>Our model was pretty good! In fact it was able to yield an accuracy score of 100%.</p>
<p>For the confusion matrix, the first element of the of the first row of the confusion matrix denotes the true negatives meaning the number of negative instances (denied applications) predicted by the model correctly. And the last element of the second row of the confusion matrix denotes the true positives meaning the number of positive instances (approved applications) predicted by the model correctly.</p>
<p>But if we hadn't got a perfect score what's to be done?. We can perform a <a href="https://machinelearningmastery.com/how-to-tune-algorithm-parameters-with-scikit-learn/">grid search</a> of the model parameters to improve the model's ability to predict credit card approvals.</p>
<p><a href="http://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LogisticRegression.html">scikit-learn's implementation of logistic regression</a> consists of different hyperparameters but we will grid search over the following two:</p>
<ul>
<li>tol</li>
<li>max_iter</li>
</ul>

In [52]:
# Import GridSearchCV
from sklearn.model_selection import GridSearchCV

# Define the grid of values for tol and max_iter
tol = [0.01, 0.001, 0.0001]
max_iter = [100, 150, 200]

# Create a dictionary where tol and max_iter are keys and the lists of their values are corresponding values
param_grid = dict(tol=tol, max_iter=max_iter)


## 12. Finding the best performing model
<p>We have defined the grid of hyperparameter values and converted them into a single dictionary format which <code>GridSearchCV()</code> expects as one of its parameters. Now, we will begin the grid search to see which values perform best.</p>
<p>We will instantiate <code>GridSearchCV()</code> with our earlier <code>logreg</code> model with all the data we have. We will also instruct <code>GridSearchCV()</code> to perform a <a href="https://www.dataschool.io/machine-learning-with-scikit-learn/">cross-validation</a> of five folds.</p>
<p>We'll end the notebook by storing the best-achieved score and the respective best parameters.</p>
<p>While building this credit card predictor, we tackled some of the most widely-known preprocessing steps such as <strong>scaling</strong>, <strong>label encoding</strong>, and <strong>missing value imputation</strong>. We finished with some <strong>machine learning</strong> to predict if a person's application for a credit card would get approved or not given some information about that person.</p>

In [53]:
# Instantiate GridSearchCV with the required parameters
grid_model = GridSearchCV(estimator=logreg, param_grid=param_grid, cv=5)

# Fit grid_model to the data
grid_model_result = grid_model.fit(rescaledX_train, y_train)

# Summarize results
best_score = grid_model_result.best_score_
best_params = grid_model_result.best_params_
print("Best: %f using %s" % (best_score, best_params))

# Extract the best model and evaluate it on the test set
best_model = grid_model_result.best_estimator_
accuracy_best_model = best_model.score(rescaledX_test, y_test)
print("Accuracy of the best logistic regression classifier: ", accuracy_best_model)


Best: 1.000000 using {'max_iter': 100, 'tol': 0.01}
Accuracy of the best logistic regression classifier:  1.0
